[FSDP2] Added explicit `unshard(async_op)` API

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #120952

This PR adds an `unshard(async_op: bool = False)` API to manually unshard the parameters via all-gather. This can be used for reordering the all-gather with other collectives (e.g. all-to-all).

This currently requires the user to set `TORCH_NCCL_AVOID_RECORD_STREAMS=1` to avoid `recordStream` from `ProcessGroupNCCL` and get expected memory behaviors.


cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang

Differential Revision: [D56148725](https://our.internmc.facebook.com/intern/diff/D56148725)

Commit:
[FSDP2] Added explicit `unshard(async_op)` API

[ghstack-poisoned]

Commit:
Update on "[FSDP2] Added explicit `unshard(async_op)` API"


To-do:
- Add unit testing!

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

Commit:
Update on "[FSDP2] Added explicit `unshard(async_op)` API"


This PR adds an `unshard(async_op: bool = False)` API to manually unshard the parameters via all-gather. This can be used for reordering the all-gather with other collectives (e.g. all-to-all).

To-do:
- Add unit testing!

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

Commit:
Update on "[FSDP2] Added explicit `unshard(async_op)` API"


This PR adds an `unshard(async_op: bool = False)` API to manually unshard the parameters via all-gather. This can be used for reordering the all-gather with other collectives (e.g. all-to-all).

To-do:
- Add unit testing!

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

Commit:
Update on "[FSDP2] Added explicit `unshard(async_op)` API"


This PR adds an `unshard(async_op: bool = False)` API to manually unshard the parameters via all-gather. This can be used for reordering the all-gather with other collectives (e.g. all-to-all).

To-do:
- Add unit testing!

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

Commit:
Update on "[FSDP2] Added explicit `unshard(async_op)` API"


This PR adds an `unshard(async_op: bool = False)` API to manually unshard the parameters via all-gather. This can be used for reordering the all-gather with other collectives (e.g. all-to-all).

To-do:
- Add unit testing!

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

Commit:
Update on "[FSDP2] Added explicit `unshard(async_op)` API"


This PR adds an `unshard(async_op: bool = False)` API to manually unshard the parameters via all-gather. This can be used for reordering the all-gather with other collectives (e.g. all-to-all).

To-do:
- Add unit testing!

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang

[ghstack-poisoned]

Commit:
Update on "[FSDP2] Added explicit `unshard(async_op)` API"


This PR adds an `unshard(async_op: bool = False)` API to manually unshard the parameters via all-gather. This can be used for reordering the all-gather with other collectives (e.g. all-to-all).

To-do:
- Add unit testing!

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang

[ghstack-poisoned]

Commit:
Update on "[FSDP2] Added explicit `unshard(async_op)` API"


This PR adds an `unshard(async_op: bool = False)` API to manually unshard the parameters via all-gather. This can be used for reordering the all-gather with other collectives (e.g. all-to-all).

To-do:
- Add unit testing!

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang

[ghstack-poisoned]

Commit:
Update on "[FSDP2] Added explicit `unshard(async_op)` API"


This PR adds an `unshard(async_op: bool = False)` API to manually unshard the parameters via all-gather. This can be used for reordering the all-gather with other collectives (e.g. all-to-all).

To-do:
- Add unit testing!
- Rationalize `recordStream` for `async_op=True`

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang

[ghstack-poisoned]

Commit:
Update on "[FSDP2] Added explicit `unshard(async_op)` API"


This PR adds an `unshard(async_op: bool = False)` API to manually unshard the parameters via all-gather. This can be used for reordering the all-gather with other collectives (e.g. all-to-all).

To-do:
- Add unit testing!
- Rationalize `recordStream` for `async_op=True`

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang

[ghstack-poisoned]

Commit:
Update on "[FSDP2] Added explicit `unshard(async_op)` API"


This PR adds an `unshard(async_op: bool = False)` API to manually unshard the parameters via all-gather. This can be used for reordering the all-gather with other collectives (e.g. all-to-all).

To-do:
- Add unit testing!
- Rationalize `recordStream` for `async_op=True`

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang

[ghstack-poisoned]

Commit:
Update on "[FSDP2] Added explicit `unshard(async_op)` API"


This PR adds an `unshard(async_op: bool = False)` API to manually unshard the parameters via all-gather. This can be used for reordering the all-gather with other collectives (e.g. all-to-all).

To-do:
- Add unit testing!
- Rationalize `recordStream` for `async_op=True`

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang

[ghstack-poisoned]

Commit:
Update on "[FSDP2] Added explicit `unshard(async_op)` API"


This PR adds an `unshard(async_op: bool = False)` API to manually unshard the parameters via all-gather. This can be used for reordering the all-gather with other collectives (e.g. all-to-all).

To-do:
- Add unit testing!
- Rationalize `recordStream` for `async_op=True`

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang

[ghstack-poisoned]

Commit:
Update on "[FSDP2] Added explicit `unshard(async_op)` API"


This PR adds an `unshard(async_op: bool = False)` API to manually unshard the parameters via all-gather. This can be used for reordering the all-gather with other collectives (e.g. all-to-all).

To-do:
- Add unit testing!
- Rationalize `recordStream` for `async_op=True`

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang

[ghstack-poisoned]

Commit:
Update on "[FSDP2] Added explicit `unshard(async_op)` API"


This PR adds an `unshard(async_op: bool = False)` API to manually unshard the parameters via all-gather. This can be used for reordering the all-gather with other collectives (e.g. all-to-all).

To-do:
- Add unit testing!
- Rationalize `recordStream` for `async_op=True`

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang

[ghstack-poisoned]

Commit:
Update on "[FSDP2] Added explicit `unshard(async_op)` API"


This PR adds an `unshard(async_op: bool = False)` API to manually unshard the parameters via all-gather. This can be used for reordering the all-gather with other collectives (e.g. all-to-all).

To-do:
- Add unit testing!
- Rationalize `recordStream` for `async_op=True`

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang

[ghstack-poisoned]

Commit:
Update on "[FSDP2] Added explicit `unshard(async_op)` API"


This PR adds an `unshard(async_op: bool = False)` API to manually unshard the parameters via all-gather. This can be used for reordering the all-gather with other collectives (e.g. all-to-all).

To-do:
- Add unit testing!
- Rationalize `recordStream` for `async_op=True`

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang

[ghstack-poisoned]

Commit:
Update on "[FSDP2] Added explicit `unshard(async_op)` API"


This PR adds an `unshard(async_op: bool = False)` API to manually unshard the parameters via all-gather. This can be used for reordering the all-gather with other collectives (e.g. all-to-all).

To-do:
- Add unit testing!
- Rationalize `recordStream` for `async_op=True`

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang

[ghstack-poisoned]

Commit:
Update on "[FSDP2] Added explicit `unshard(async_op)` API"


This PR adds an `unshard(async_op: bool = False)` API to manually unshard the parameters via all-gather. This can be used for reordering the all-gather with other collectives (e.g. all-to-all).

To-do:
- Add unit testing!
- Rationalize `recordStream` for `async_op=True`

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang

[ghstack-poisoned]

awgu commented on test/distributed/_composable/fsdp/test_fully_shard_comm.py:
```diff
@@ -573,5 +577,91 @@ def post_backward_with_record(self, *args, **kwargs):
+        fully_shard(model.mlps)
+        replicate(model.cuda())
+        optim = torch.optim.Adam(model.parameters(), lr=1e-2, foreach=True)
+        if self.rank == 0:
+            print(model)
```
To-do: remove

Commit:
Update on "[FSDP2] Added explicit `unshard(async_op)` API"


This PR adds an `unshard(async_op: bool = False)` API to manually unshard the parameters via all-gather. This can be used for reordering the all-gather with other collectives (e.g. all-to-all).

This currently requires the user to set `TORCH_NCCL_AVOID_RECORD_STREAMS=1` to avoid `recordStream` from `ProcessGroupNCCL` and get expected memory behaviors.


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang

[ghstack-poisoned]

Commit:
Update on "[FSDP2] Added explicit `unshard(async_op)` API"


This PR adds an `unshard(async_op: bool = False)` API to manually unshard the parameters via all-gather. This can be used for reordering the all-gather with other collectives (e.g. all-to-all).

This currently requires the user to set `TORCH_NCCL_AVOID_RECORD_STREAMS=1` to avoid `recordStream` from `ProcessGroupNCCL` and get expected memory behaviors.


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang

[ghstack-poisoned]

Commit:
Update on "[FSDP2] Added explicit `unshard(async_op)` API"


This PR adds an `unshard(async_op: bool = False)` API to manually unshard the parameters via all-gather. This can be used for reordering the all-gather with other collectives (e.g. all-to-all).

This currently requires the user to set `TORCH_NCCL_AVOID_RECORD_STREAMS=1` to avoid `recordStream` from `ProcessGroupNCCL` and get expected memory behaviors.


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang

[ghstack-poisoned]

weifengpy commented on test/distributed/_composable/fsdp/test_fully_shard_comm.py:
```diff
@@ -573,5 +577,89 @@ def post_backward_with_record(self, *args, **kwargs):
+        return min(torch.cuda.device_count(), 2)
+
+    @skip_if_lt_x_gpu(2)
+    @wrapSwapTensorsTest(True)
```
when shoudl we use `@wrapSwapTensorsTest(True)` ?

awgu replied:
Good catch! We do not need this anymore since a PR has landed in core now that defaults to use `swap_tensors` for wrapper subclasses like `DTensor`. This is a remnant from before and should be removed.

Commit:
Update on "[FSDP2] Added explicit `unshard(async_op)` API"


This PR adds an `unshard(async_op: bool = False)` API to manually unshard the parameters via all-gather. This can be used for reordering the all-gather with other collectives (e.g. all-to-all).

This currently requires the user to set `TORCH_NCCL_AVOID_RECORD_STREAMS=1` to avoid `recordStream` from `ProcessGroupNCCL` and get expected memory behaviors.


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang

[ghstack-poisoned]

Commit:
Update on "[FSDP2] Added explicit `unshard(async_op)` API"


This PR adds an `unshard(async_op: bool = False)` API to manually unshard the parameters via all-gather. This can be used for reordering the all-gather with other collectives (e.g. all-to-all).

This currently requires the user to set `TORCH_NCCL_AVOID_RECORD_STREAMS=1` to avoid `recordStream` from `ProcessGroupNCCL` and get expected memory behaviors.


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D56148725).

wanchaol commented on torch/distributed/_composable/fsdp/fully_shard.py:
```diff
@@ -161,6 +161,33 @@ def reshard(self) -> None:
+                this function.
+
+        .. note:: If ``async_op=True``, then the user does not have to call
+            :meth:`wait` on the returned handle if the unshard op can be waited
```
> if the unshard op can be waited on in the module's pre-forward.

wouldn't this always the case? the module's pre-forward would always wait on the pending unshard op on this module itself?

awgu replied:
Sorry, I meant more like "if waiting on the unshard op in the module's pre-forward _is tolerable_".

There are two reasons why a user may not want to wait until the pre-forward to naturally call `.wait()` / `wait_for_unshard()`:
1. The user wants to use the result of the all-gather earlier (unlikely)
2. If we use PG stashing (i.e. `TORCH_NCCL_AVOID_RECORD_STREAMS=1`), then calling `.wait()` allows the all-gather input to be freed, which saves memory -- the user may want to call `.wait()` earlier to free the memory earlier

awgu replied:
added some clarification in the doc

wanchaol commented on torch/distributed/_composable/fsdp/fully_shard.py:
```diff
@@ -247,3 +274,14 @@ def _apply(self, *args: Any, **kwargs: Any) -> Any:
         return ret
+
+
+class UnshardHandle:
```
may need to add some docs on this class since it's a public API?

awgu replied:
I think docs are not rendered yet since this is in a private folder, but let me add some docs.

wanchaol reviewed (approved):
looks good! have only one question

Commit:
Update on "[FSDP2] Added explicit `unshard(async_op)` API"


This PR adds an `unshard(async_op: bool = False)` API to manually unshard the parameters via all-gather. This can be used for reordering the all-gather with other collectives (e.g. all-to-all).

This currently requires the user to set `TORCH_NCCL_AVOID_RECORD_STREAMS=1` to avoid `recordStream` from `ProcessGroupNCCL` and get expected memory behaviors.


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang

Differential Revision: [D56148725](https://our.internmc.facebook.com/intern/diff/D56148725)

[ghstack-poisoned]

awgu commented:
@awgu has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D56148725).

awgu commented:
@pytorchbot merge

weifengpy commented:
I find this unshard API to be helpful for debugging. for exmample, I can inspect full params outside fwd/bwd. convenient!
```
loss = self._loss_fn(logits, labels)
self._model.unshard()
self._model.layers[0].unshard()
# inspect params
```
